plugins/grpc/stats/client_handler.go

```diff
@@ -0,0 +1,176 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package stats
+
+import (
+	"context"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	istats "github.com/census-instrumentation/opencensus-go/stats"
+	"github.com/census-instrumentation/opencensus-go/tags"
+	"github.com/golang/glog"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+)
+
+var (
+	// grpcClientConnKey is the key used to store client instrumentation
+	// connection related data into the context.
+	grpcClientConnKey *grpcInstrumentationKey
+	// grpcClientRPCKey is the key used to store client instrumentation RPC
+	// related data into the context.
+	grpcClientRPCKey *grpcInstrumentationKey
+)
+
+// ClientHandler is the type implementing the "google.golang.org/grpc/stats.Handler"
+// interface to process lifecycle events from the GRPC client.
+type ClientHandler struct{}
+
+// TagConn adds connection related data to the given context and returns the
+// new context.
+func (ch ClientHandler) TagConn(ctx context.Context, info *stats.ConnTagInfo) context.Context {
+	// Do nothing. This is here to satisfy the interface "google.golang.org/grpc/stats.Handler"
+	return ctx
+}
+
+// HandleConn processes the connection events.
+func (ch ClientHandler) HandleConn(ctx context.Context, s stats.ConnStats) {
+	// Do nothing. This is here to satisfy the interface "google.golang.org/grpc/stats.Handler"
+}
+
+// TagRPC gets the github.com/census-instrumentation/opencensus-go/tags.TagsSet
+// populated by the application code, serializes its tags into the GRPC
+// metadata in order to be sent to the server.
+func (ch ClientHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	startTime := time.Now()
+	if ctx == nil {
+		if glog.V(2) {
+			glog.Infoln("ClientHandler.TagRPC called with nil context")
+		}
+		return ctx
+	}
+
+	if info == nil {
+		if glog.V(2) {
+			glog.Infof("ClientHandler.TagRPC called with nil info.", info.FullMethodName)
+		}
+		return ctx
+	}
+	names := strings.Split(info.FullMethodName, "/")
+	if len(names) != 3 {
+		if glog.V(2) {
+			glog.Infof("ClientHandler.TagRPC called with info.FullMethodName bad format. got %v, want '/$service/$method/'", info.FullMethodName)
+		}
+		return ctx
+	}
+	serviceName := names[1]
+	methodName := names[2]
+
+	d := &rpcData{
+		startTime: startTime,
+	}
+
+	ts := tags.FromContext(ctx)
+	encoded := tags.EncodeToFullSignature(ts)
+
+	// Join old metadata with new one, so the old metadata information is not lost.
+	md, _ := metadata.FromOutgoingContext(ctx)
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Join(md, metadata.Pairs(tagsKey, string(encoded))))
+
+	tsb := tags.NewTagSetBuilder(ts)
+	tsb.UpsertString(keyService, serviceName)
```

Comment:
Why does this function return a `TagSetBuilder` if it's called to update the builder?

Reply:
To make it easier to use the shortcut version when modifying multiple tags:
tags.NewTagSetBuilder(ts).
        UpsertString(keyService, serviceName).
        UpsertString(key2, value2).
        AddString(key3, value3)...